enterprise_gateway/services/kernelspecs/handlers.py

```diff
@@ -1,19 +1,117 @@
 # Copyright (c) Jupyter Development Team.
 # Distributed under the terms of the Modified BSD License.
 """Tornado handlers for kernel specs."""
-
-import notebook.services.kernelspecs.handlers as notebook_handlers
+import json
 import notebook.kernelspecs.handlers as notebook_kernelspecs_resources_handlers
-
+from notebook.services.kernelspecs.handlers import is_kernelspec_model, kernelspec_model
+from notebook.utils import maybe_future, url_unescape
+from tornado import web
+from ...base.handlers import APIHandler
 from ...mixins import TokenAuthorizationMixin, CORSMixin, JSONErrorsMixin
 
+
+def key_exists(obj, chain):
+    _key = chain.pop(0)
+    if _key in obj:
+        return key_exists(obj[_key], chain) if chain else obj[_key]
+
+
+def apply_user_filter(kernelspec_model, kernel_user=None):
+    if kernel_user:
+        # Check unauthorized list
+        if key_exists(kernelspec_model, ['spec', 'metadata', 'process_proxy', 'config', 'unauthorized_users']):
+            # Check if kernel_user in kernelspec_model
+            unauthorized_list = kernelspec_model['spec']['metadata']['process_proxy']['config']['unauthorized_users']
+            if kernel_user in unauthorized_list:
+                return None
+        # Check authorized list
+        if key_exists(kernelspec_model, ['spec', 'metadata', 'process_proxy', 'config', 'authorized_users']):
+            authorized_list = kernelspec_model['spec']['metadata']['process_proxy']['config']['authorized_users']
+            if authorized_list and kernel_user not in authorized_list:
+                return None
+    return kernelspec_model
+
+
+class MainKernelSpecHandler(TokenAuthorizationMixin,
+                            CORSMixin,
+                            JSONErrorsMixin,
+                            APIHandler):
+    @web.authenticated
+    async def get(self):
+        ksm = self.kernel_spec_manager
+        km = self.kernel_manager
+        model = {}
+        model['default'] = km.default_kernel_name
+        model['kernelspecs'] = specs = {}
+
+        kernel_user_filter = self.request.query_arguments.get('user')
+        kernel_user = None
+        if kernel_user_filter:
+            kernel_user = kernel_user_filter[0].decode("utf-8")
+            if kernel_user:
+                self.log.debug("Searching kernels for user '%s' " % kernel_user)
+
+        kspecs = await maybe_future(ksm.get_all_specs())
+
+        list_kernels_found = []
+        for kernel_name, kernel_info in kspecs.items():
+            try:
+                if is_kernelspec_model(kernel_info):
+                    d = kernel_info
+                else:
+                    d = kernelspec_model(self, kernel_name, kernel_info['spec'], kernel_info['resource_dir'])
+                d = apply_user_filter(d, kernel_user)
+                if d is not None:
+                    specs[kernel_name] = d
+                    list_kernels_found.append(d['name'])
```

Comment:
I think it would be good to add a DEBUG message in an `else` block for the unauthorized case.  This is because when debug is enabled, EG will produce the following kind of output...
```
[D 2020-04-28 11:55:46.982 EnterpriseGatewayApp] Searching kernels for user 'scala' 
[D 2020-04-28 11:55:46.983 EnterpriseGatewayApp] Found kernel python3 in /opt/anaconda2/envs/py3/share/jupyter/kernels
[D 2020-04-28 11:55:46.983 EnterpriseGatewayApp] Found kernel ir in /opt/anaconda2/envs/py3/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel dask_python_yarn_remote in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_r_yarn_client in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_r_yarn_cluster in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_python_yarn_client in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_python_yarn_cluster in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_scala_yarn_client in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_scala_yarn_cluster in /usr/local/share/jupyter/kernels
```
In my example, user `scala` is in the unauthorized users list for the python-related YARN kernels and the proper filtering is applied.  However, because the "Found kernel" message is produced from `jupyter_client` this output can be misleading.

I think a DEBUG statement along the lines of: "User '%s' is not authorized to use kernel spec %s, excluding from results." would be helpful. 

Here's what the above would then look like...
```
[D 2020-04-28 11:55:46.982 EnterpriseGatewayApp] Searching kernels for user 'scala' 
[D 2020-04-28 11:55:46.983 EnterpriseGatewayApp] Found kernel python3 in /opt/anaconda2/envs/py3/share/jupyter/kernels
[D 2020-04-28 11:55:46.983 EnterpriseGatewayApp] Found kernel ir in /opt/anaconda2/envs/py3/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel dask_python_yarn_remote in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] User 'scala' is not authorized to use kernel spec dask_python_yarn_remote, excluding from results.
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_r_yarn_client in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_r_yarn_cluster in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_python_yarn_client in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] User 'scala' is not authorized to use kernel spec spark_python_yarn_client, excluding from results.
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_python_yarn_cluster in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] User 'scala' is not authorized to use kernel spec spark_python_yarn_cluster, excluding from results.
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_scala_yarn_client in /usr/local/share/jupyter/kernels
[D 2020-04-28 11:55:46.984 EnterpriseGatewayApp] Found kernel spark_scala_yarn_cluster in /usr/local/share/jupyter/kernels
```